### PR TITLE
use generics in random

### PR DESCRIPTION
**Pull request by ringabout**
Ref https://github.com/timotheecour/Nim/issues/408

### CODE REVIEW

File: lib/pure/random.nim [hunk 1]
```diff
@@ -233,28 +232,6 @@ proc rand*(r: var Rand; max: Natural): int {.benign.} =
...
       return int(x mod (uint64(max)+1u64))
 
-proc rand*(max: int): int {.benign.} =
-  ## Returns a random integer in the range `0..max`.
-  ##
-  ## If `randomize<#randomize>`_ has not been called, the sequence of random
-  ## numbers returned from this proc will always be the same.
-  ##
-  ## This proc uses the default random number generator. Thus, it is **not**
-  ## thread-safe.
```
Comment (Araq) [inline lib/pure/random.nim:243]: What about this remark? Shouldn't we keep it?

Comment (ringabout) [reply lib/pure/random.nim:243]: It is already kept
https://github.com/nim-lang/Nim/blob/de29bed82e40597a5d6450e697f06a9f9f87d16a/lib/pure/random.nim#L264

`rand(max: float)` already has the same comments and this generic procs reuse that commentbefore.

File: lib/pure/random.nim [hunk 2]
```diff
@@ -276,8 +253,10 @@ proc rand*(r: var Rand; max: range[0.0 .. high(float)]): float {.benign.} =
     let u = (0x3FFu64 shl 52u64) or (x shr 12u64)
     result = (cast[float](u) - 1.0) * max
 
-proc rand*(max: float): float {.benign.} =
-  ## Returns a random floating point number in the range `0.0..max`.
+proc rand*[T: Ordinal or SomeFloat](max: T): T {.benign.} =
+  ## Returns a random floating point number in the range `T(0) .. max`.
+  ## 
+  ## Allowed types for `T` are integers, floats, and enums without holes.
```
Comment (timotheecour) [inline lib/pure/random.nim:259]: this comment is redundant (and only approximately correct) and should be removed, it's already specified what T is in the generated docs:

image

furthermore, `Ordinal` is clickable (thanks to docgen) so users who see `Ordinal` for the 1st time can just click to see the definition, which is explained in 1 place, where it belongs (https://nim-lang.github.io/Nim/system.html#Ordinal) => "1 definition rule" for docs

note that the definition for https://nim-lang.github.io/Nim/system.html#Ordinal should be fixed via `s/enums/enums without holes/` but that's a separate issue

File: lib/pure/random.nim [hunk 3]
```diff
@@ -276,8 +253,10 @@ proc rand*(r: var Rand; max: range[0.0 .. high(float)]): float {.benign.} =
     let u = (0x3FFu64 shl 52u64) or (x shr 12u64)
     result = (cast[float](u) - 1.0) * max
 
-proc rand*(max: float): float {.benign.} =
-  ## Returns a random floating point number in the range `0.0..max`.
+proc rand*[T: Ordinal or SomeFloat](max: T): T {.benign.} =
+  ## Returns a random floating point number in the range `T(0) .. max`.
```
Comment (timotheecour) [inline lib/pure/random.nim:257]: s/random floating point number/random number/

**Comment by timotheecour** [comment]
@xflywind thanks for tackling this; but this needs more care:
```nim
when defined case1:
  import std/random
  when defined case1a:
    # Error: type mismatch: got <Rand, Foo>
    type Fooa = enum k0,k1,k2
    echo rand(Fooa.high)
  when defined case1b:
    # Error: type mismatch: got <Rand, Dollar>
    type Dollar = distinct int
    echo rand(int.high.Dollar)
  when defined case1c: # ok
    echo rand(int64.high)
  when defined case1d:
    # Error: unhandled exception: value out of range [RangeDefect]
    echo rand(uint64.high)
  when defined case1e:
    # Error: type mismatch: got <Rand, char>
    echo (rand(char.high),)
```

in particular this should also be updated to be made generic:
```nim
proc rand*(r: var Rand; max: Natural): int {.benign.} =
```
and you can move the old underlying `proc rand*(r: var Rand; max: Natural): int {.benign.} =` implementation to some 
```nim
proc randImpl(r: var Rand; max: Natural): int {.benign.} = ...
```

File: lib/pure/random.nim [hunk 4]
```diff
@@ -286,16 +265,21 @@ proc rand*(max: float): float {.benign.} =
...
-    randomize(234)
+    randomize(123)
+    doAssert rand(100) == 0
+    doAssert rand(100) == 96
+    doAssert rand(100) == 66
+
     let f = rand(1.0)
     ## f = 8.717181376738381e-07
-  rand(state, max)
+  result = T(rand(state, max))
```
Comment (timotheecour) [inline lib/pure/random.nim:282]: cast to avoid overflow errors

**Pull request merged**